src/lib.rs

```diff
@@ -1,14 +1,180 @@
-        assert_eq!(result, 4);
+struct AggregatedQueries {
+    pattern_index_lookup: Vec<(RuleIndex, RuleListenerIndex)>,
+    #[allow(dead_code)]
+    query: Query,
+    query_text: String,
+}
+
+impl AggregatedQueries {
+    pub fn new(resolved_rules: &[ResolvedRule], context: &Context) -> Self {
+        let mut pattern_index_lookup: Vec<(RuleIndex, RuleListenerIndex)> = Default::default();
+        let mut aggregated_query_text = String::new();
+        for (rule_index, resolved_rule) in resolved_rules.into_iter().enumerate() {
+            for (rule_listener_index, rule_listener) in resolved_rule.listeners.iter().enumerate() {
+                for _ in 0..rule_listener.query.pattern_count() {
+                    pattern_index_lookup.push((rule_index, rule_listener_index));
+                }
+                let use_capture_name =
+                    &rule_listener.query.capture_names()[rule_listener.capture_index as usize];
+                let query_text_with_unified_capture_name =
+                    regex!(&format!(r#"@{use_capture_name}\b"#)).replace_all(
+                        &rule_listener.query_text,
+                        CAPTURE_NAME_FOR_TREE_SITTER_GREP_WITH_LEADING_AT,
+                    );
+                assert!(
+                    matches!(query_text_with_unified_capture_name, Cow::Owned(_),),
+                    "Didn't find any instances of the capture name to replace"
+                );
+                aggregated_query_text.push_str(&query_text_with_unified_capture_name);
+                aggregated_query_text.push_str("\n\n");
+            }
+        }
+        let query = Query::new(context.language, &aggregated_query_text).unwrap();
+        assert!(query.pattern_count() == pattern_index_lookup.len());
+        Self {
+            pattern_index_lookup,
+            query,
+            query_text: aggregated_query_text,
+        }
     }
 }
+
+fn get_rules() -> Vec<Rule> {
+    vec![no_default_default_rule(), no_lazy_static_rule()]
+}
+
+fn no_default_default_rule() -> Rule {
+    RuleBuilder::default()
+        .name("no_default_default")
+        .create(|_context| {
+            vec![RuleListenerBuilder::default()
+                .query(
+                    r#"(
+                      (call_expression
+                        function:
+                          (scoped_identifier
+                            path:
+                              (identifier) @first (#eq? @first "Default")
+                            name:
+                              (identifier) @second (#eq? @second "default")
+                          )
+                      ) @c
+                    )"#,
+                )
+                .capture_name("c")
+                .on_query_match(|node, query_match_context| {
+                    query_match_context.report(
+                        ViolationBuilder::default()
+                            .message(r#"Use '_d()' instead of 'Default::default()'"#)
```

Comment:
These are actual rules I thought of for the Typescript compiler in Rust code-base

Specifically here I've been playing with using a "shorthand" `_d()` helper vs the more verbose `Default::default()`

Neither of them are "flexing" the tree-sitter side of it very hard but whatever

Reply:
These read pretty nicely